### PR TITLE
Add scripts and actions for creating Windows containers for Windows 10 and 11.

### PR DESCRIPTION
**Pull request by wmaxey**
This makes 2 images for each compiler and CUDA entry in the matrix of:

2017, 2019, and 2022.
CUDA 11.1 and 12.1.

See packages here for results:
https://github.com/wmaxey?tab=packages&repo_name=cuda-windows-images

### CODE REVIEW

File: .github/workflows/trigger_windows.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+jobs:
+  build-all-images:
+    name: Build Windows images
+    strategy:
+      fail-fast: false
+      matrix:
+        runner: [windows-2019, windows-2022]
+        cuda_version: ["12.1", "11.1"]
+        edition: [windows-2019, windows-2022]
```
Comment (wmaxey) [inline .github/workflows/trigger_windows.yml:21]: Runner/edition can be different in the future. We'll likely refactor this before a new edition of GHA runners comes out.

**Review by trxcllnt — APPROVED** [review]
LGTM. Should this be integrated into the main release job, or is it fine if we manually trigger releases separately?

**Comment by wmaxey** [comment]
> LGTM. Should this be integrated into the main release job, or is it fine if we manually trigger releases separately?

I think manually is fine, or at the least your choice. Pushing the tags to dockerhub wasn't included in these changes.

**Comment by trxcllnt** [comment]
Ok, we'll probably want to do that.

**Comment by jrhemstad** [comment]
@miscco

**Comment by miscco** [comment]
@trxcllnt I have made some minor modifications to the PR. With those I was able to locally build the images and run our CI scripts inside the container.

**Comment by wmaxey** [comment]
> @trxcllnt I have made some minor modifications to the PR. With those I was able to locally build the images and run our CI scripts inside the container.

I unfortunately rebased and force pushed over everything. I've updated to 12.2 and fixed the PR thing so it actually runs jobs.

**Comment by wmaxey** [comment]
Can publish images now only when desired by launching manually.

File: .github/workflows/dispatch-windows.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        description: Enable pushing images
+
+jobs:
+  build-all-images:
+    name: Build Windows images
+    strategy:
+      fail-fast: false
+      matrix:
+        runner: [windows-2019, windows-2022]
+        cuda_version: ["12.1", "11.1"]
```
Comment (trxcllnt) [inline .github/workflows/dispatch-windows.yml:18]: ?
```suggestion
        cuda_version: ["12.2", "11.1"]
```

File: scripts/windows/make-image-manifest.ps1 [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }
+}
+
+Push-location "$PSScriptRoot"
+
+try {
+    $image_name_2019="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition "windows2019" -repo $repo)"
+    $image_name_2022="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition "windows2022" -repo $repo)"
+
+    $manifest_name="${repo}:cuda${cudaVersion}-cl${clVersion}"
```
Comment (trxcllnt) [inline scripts/windows/make-image-manifest.ps1:26]:
```suggestion
    $manifest_name="${repo}:${prefix}cuda${cudaVersion}-cl${clVersion}"
```

File: scripts/windows/make-image-manifest.ps1 [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+Param(
+    [Parameter(Mandatory=$true)]
+    [string]
+    $clVersion="latest",
+    [Parameter(Mandatory=$false)]
+    [string]
+    $cudaVersion="latest",
+    [Parameter(Mandatory=$false)]
+    [string]
+    $repo="local"
```
Comment (trxcllnt) [inline scripts/windows/make-image-manifest.ps1:11]:
```suggestion
    $repo="local",
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

File: .github/actions/create-and-push-manifest/action.ps1 [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Param(
+    [Parameter(Mandatory=$true)]
+    [string]
+    $msvcVersion,
+    [Parameter(Mandatory=$true)]
+    [string]
+    $cudaVersion="latest",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $repo
```
Comment (trxcllnt) [inline .github/actions/create-and-push-manifest/action.ps1:10]:
```suggestion
    $repo,
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

File: .github/actions/create-and-push-manifest/action.ps1 [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+$ErrorActionPreference = "Stop"
+
+# Assume this script is launched from repo root.
+.\scripts\windows\vs-version-matrix.ps1
+$clVerArray = $vsVerToCompilers[$msvcVersion]
+
+foreach ($cl in $clVerArray) {
+    .\scripts\windows\make-image-manifest -clVersion $cl -cudaVersion $cudaVersion -repo $repo
```
Comment (trxcllnt) [inline .github/actions/create-and-push-manifest/action.ps1:20]:
```suggestion
    .\scripts\windows\make-image-manifest -clVersion $cl -cudaVersion $cudaVersion -repo $repo --prefix $prefix
```

File: .github/actions/create-and-push-manifest/action.yml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  using: composite
+  steps:
+    - name: Run action script to build images
+      id: build
+      shell: powershell
+      run: |
+        .github\actions\create-and-push-manifest\action.ps1 `
+          -cudaVersion ${{ inputs.cuda_version }}           `
+          -msvcVersion ${{ inputs.vs_version }}             `
+          -repo ${{ inputs.repo }}
```
Comment (trxcllnt) [inline .github/actions/create-and-push-manifest/action.yml:28]: Not sure about powershell syntax, but something like this?
```suggestion
          -repo ${{ inputs.repo }}
          -prefix "$(git describe --abbrev=0 --tags | sed 's/[a-zA-Z]//g' | cut -d '.' -f -2)-"
```

File: scripts/windows/build-windows-image.ps1 [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    "windows-2019" = "mcr.microsoft.com/windows/servercore:ltsc2019"
+}[$edition]
+
+try {
+    # Source version matrix
+    .\vs-version-matrix.ps1
+
+    $vsVer = $vsYearToVer[$vsCompilersToYear[$clVersion]]
+    # Override defaults in .env.
+    $ENV:IMAGE_NAME="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition $edition -repo $repo)"
```
Comment (trxcllnt) [inline scripts/windows/build-windows-image.ps1:41]:
```suggestion
    $ENV:IMAGE_NAME="$(.\generate-image-name.ps1 -clVersion $clVersion -cudaVersion $cudaVersion -edition $edition -repo $repo -prefix $prefix)"
```

File: scripts/windows/build-windows-image.ps1 [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    [ValidateSet('windows-2019', 'windows-2022')]
+    [string]
+    $edition="windows",
+    [Parameter(Mandatory=$false)]
+    [ValidateSet('hyperv', 'process')]
+    [string]
+    $isolation="hyperv",
+    [Parameter(Mandatory=$false)]
+    [string]
+    $repo="local"
```
Comment (trxcllnt) [inline scripts/windows/build-windows-image.ps1:19]:
```suggestion
    $repo="local",
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

File: .github/actions/build-windows-image/action.ps1 [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+$ErrorActionPreference = "Stop"
+
+# Assume this script is launched from repo root.
+.\scripts\windows\vs-version-matrix.ps1
+$clVerArray = $vsVerToCompilers[$msvcVersion]
+
+foreach ($cl in $clVerArray) {
+    .\scripts\windows\build-windows-image.ps1 -clVersion $cl -isolation $isolation -cudaVersion $cudaVersion -edition $edition -repo $repo
```
Comment (trxcllnt) [inline .github/actions/build-windows-image/action.ps1:26]:
```suggestion
    .\scripts\windows\build-windows-image.ps1 -clVersion $cl -isolation $isolation -cudaVersion $cudaVersion -edition $edition -repo $repo -prefix $prefix
```

File: .github/actions/build-windows-image/action.ps1 [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    $cudaVersion="latest",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $edition="windows",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $isolation="hyperv",
+    [Parameter(Mandatory=$true)]
+    [string]
+    $repo
```
Comment (trxcllnt) [inline .github/actions/build-windows-image/action.ps1:16]:
```suggestion
    $repo,
    [Parameter(Mandatory=$false)]
    [string]
    $prefix=""
```

File: .github/actions/build-windows-image/action.yml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    - name: Run action script to build images
+      id: build
+      shell: powershell
+      run: |
+        .github/actions/build-windows-image/action.ps1 `
+          -cudaVersion ${{ inputs.cuda_version }}      `
+          -msvcVersion ${{ inputs.vs_version }}        `
+          -edition ${{ inputs.win_edition }}           `
+          -isolation ${{ inputs.isolation }}           `
+          -repo ${{ inputs.repo }}
```
Comment (trxcllnt) [inline .github/actions/build-windows-image/action.yml:38]: Not sure about powershell syntax, but something like this?
```suggestion
          -repo ${{ inputs.repo }}
          -prefix "$(git describe --abbrev=0 --tags | sed 's/[a-zA-Z]//g' | cut -d '.' -f -2)-"
```

**Comment by miscco** [comment]
@trxcllnt how is the state of the PR, can it be merged?

**Pull request merged**